### PR TITLE
[FEATURE] `argilla`: add support to distribution

### LINKED ISSUE #5033: [TASK] Support new optional `distribution` attribute to be assigned when creating datasets

### LINKED ISSUE #5034: [TASK] Support to change `distribution` settings when updating datasets

### PR DESCRIPTION
**Pull request by frascuchon**
# Description

This PR adds support to configure the task distribution strategy when creating or updating datasets.

We can create datasets with specific task distribution setup
```python

task_distribution = TaskDistribution(min_submitted=4)

settings = Settings(
    fields=[TextField(name="text", title="text")],
    questions=[LabelQuestion(name="label", title="text", labels=["positive", "negative"])],
    distribution=task_distribution,
)
dataset = Dataset(dataset_name, settings=settings).create()
```

or update an existing dataset (without any user response)
```python
dataset = client.datasets(...)

dataset.settings.distribution.min_submitted = 100
# or 
dataset.distribution.min_submitted = 100
# or 
dataset.distribution = TaskDistribution(min_submitted=100)
dataset.update()
```

Closes https://github.com/argilla-io/argilla/issues/5033
Closes https://github.com/argilla-io/argilla/issues/5034

Refs: https://github.com/argilla-io/argilla/pull/5246

**Type of change**

- New feature (non-breaking change which adds functionality)
- Improvement (change adding some improvement to an existing functionality)

**How Has This Been Tested**

**Checklist**

- I added relevant documentation
- I followed the style guidelines of this project
- I did a self-review of my code
- I made corresponding changes to the documentation
- I confirm My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature works
- I have added relevant notes to the CHANGELOG.md file (See https://keepachangelog.com/)

### CODE REVIEW

File: argilla/src/argilla/_api/_datasets.py [hunk 1]
```diff
@@ -35,7 +35,7 @@ class DatasetsAPI(ResourceAPI[DatasetModel]):
 
     @api_error_handler
     def create(self, dataset: "DatasetModel") -> "DatasetModel":
-        json_body = dataset.model_dump()
+        json_body = dataset.model_dump(exclude_unset=True)
```
Comment (frascuchon) [inline argilla/src/argilla/_api/_datasets.py:38]: We ONLY should send what the user provided:

1. This will let the server set their defaults (when no data was sent)
2. Avoids sending `None` which may result in wrong inputs

File: .github/workflows/argilla.yml [hunk 2]
```diff
@@ -21,7 +21,7 @@ jobs:
   build:
     services:
       argilla-quickstart:
-        image: argilla/argilla-quickstart:main
+        image: argilladev/argilla-quickstart:pr-5136
```
Comment (frascuchon) [inline .github/workflows/argilla.yml]: Using the feat. branch quickstart image including all the backend changes

Comment (jfcalvo) [reply .github/workflows/argilla.yml]: I guess we need to revert this after merging to `develop`.

File: argilla/src/argilla/_api/_datasets.py (hunk 1 above)
Comment (jfcalvo) [reply argilla/src/argilla/_api/_datasets.py:38]: It is necessary to review the rest of resources for this?

File: argilla/src/argilla/_models/_dataset.py [hunk 3]
```diff
@@ -31,9 +33,10 @@ class DatasetModel(ResourceModel):
     guidelines: Optional[str] = None
     allow_extra_metadata: bool = True  # Ideally, the default value should be provided by the server
 
+    distribution: Optional[TaskDistributionModel] = None
+
     workspace_id: Optional[UUID] = None
     last_activity_at: Optional[datetime] = None
```
Comment (jfcalvo) [inline argilla/src/argilla/_models/_dataset.py]:
```suggestion
    guidelines: Optional[str] = None
    allow_extra_metadata: bool = True  # Ideally, the default value should be provided by the server
    distribution: Optional[TaskDistributionModel] = None
    workspace_id: Optional[UUID] = None
    last_activity_at: Optional[datetime] = None
```

File: argilla/src/argilla/settings/_resource.py [hunk 4]
```diff
@@ -70,6 +75,8 @@ def __init__(
         self.__guidelines = self.__process_guidelines(guidelines)
         self.__allow_extra_metadata = allow_extra_metadata
 
+        self._distribution = distribution or DEFAULT_TASK_DISTRIBUTION
```
Comment (jfcalvo) [inline argilla/src/argilla/settings/_resource.py:78]: Is not better to default to `None` and left the backend be the one selecting the default value? I guess `exclude_unset` is required then.

File: .github/workflows/argilla.yml (hunk 2 above)
Comment (davidberenstein1957) [reply .github/workflows/argilla.yml]: this should be changed back right?

File: argilla/src/argilla/_api/_datasets.py [hunk 5]
```diff
@@ -48,13 +48,13 @@ def create(self, dataset: "DatasetModel") -> "DatasetModel":
...
     def update(self, dataset: "DatasetModel") -> "DatasetModel":
-        json_body = dataset.model_dump()
+        json_body = dataset.model_dump(exclude_unset=True)
         dataset_id = json_body["id"]  # type: ignore
         response = self.http_client.patch(f"{self.url_stub}/{dataset_id}", json=json_body)
         response.raise_for_status()
         response_json = response.json()
         dataset = self._model_from_json(response_json=response_json)
-        self._log_message(message=f"Updated dataset {dataset.url}")
+        self._log_message(message=f"Updated dataset {dataset.id}")
```
Comment (davidberenstein1957) [inline argilla/src/argilla/_api/_datasets.py:57]: i would say the url is more useful for humans. why did you change this?

File: argilla/src/argilla/_api/_datasets.py [hunk 6]
```diff
@@ -63,7 +63,7 @@ def get(self, dataset_id: UUID) -> "DatasetModel":
         response.raise_for_status()
         response_json = response.json()
         dataset = self._model_from_json(response_json=response_json)
-        self._log_message(message=f"Got dataset {dataset.url}")
+        self._log_message(message=f"Got dataset {dataset.id}")
```
Comment (davidberenstein1957) [inline argilla/src/argilla/_api/_datasets.py:66]: same here

File: argilla/src/argilla/_models/_settings/_task_distribution.py [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+__all__ = ["TaskDistributionModel", "OverlapTaskDistributionModel"]
+
+from typing import Literal
+
+from pydantic import BaseModel, PositiveInt, ConfigDict
+
+
+class OverlapTaskDistributionModel(BaseModel):
+    strategy: Literal["overlap"]
+    min_submitted: PositiveInt
```
Comment (davidberenstein1957) [inline argilla/src/argilla/_models/_settings/_task_distribution.py:24]: why not n_submitted?

File: argilla/src/argilla/_api/_datasets.py (hunk 1 above)
Comment (frascuchon) [reply argilla/src/argilla/_api/_datasets.py:38]: I can review them in a separate PR

File: .github/workflows/argilla.yml (hunk 2 above)
Comment (frascuchon) [reply .github/workflows/argilla.yml]: Yes, once the branch is merged in develop, we can use the develop branch (but still the `argilladev` namespace)

File: argilla/src/argilla/settings/_resource.py (hunk 4 above)
Comment (frascuchon) [reply argilla/src/argilla/settings/_resource.py:78]: I see your point, but I think is easier for users to create settings with default values. Otherwise may bring errors due to None values

File: argilla/src/argilla/_api/_datasets.py (hunk 5 above)
Comment (frascuchon) [reply argilla/src/argilla/_api/_datasets.py:57]: Because the `url` attribute does not exist. For that, we should create it based on client settings, and this is out of the scope of this PR. We can use the dataset name for now.

File: argilla/src/argilla/_models/_settings/_task_distribution.py (hunk 7 above)
Comment (frascuchon) [reply argilla/src/argilla/_models/_settings/_task_distribution.py:24]: I feel `n_submitted` is a bit poor and can bring some confusion: max number of submitted responses? Min number? Also, `min_submitted` has been used from the beginning of discussions for this feature and I think it works well.

File: argilla/src/argilla/_api/_datasets.py (hunk 5 above)
Comment (davidberenstein1957) [reply argilla/src/argilla/_api/_datasets.py:57]: @frascuchon, ahh okay. I did not realize this was not included. I created this issues to see if anyone from the community might be interested in pciking it up. https://github.com/argilla-io/argilla/issues/5195

**Review by burtenshaw — COMMENTED** [review]
Looking at this, I realise that add 'distribution' to `Settings` and not `Dataset` could be problematic.

The settings should be attached to the feedback task definition, and the dataset should be attached to the Argilla annotation process. So a user should be able to attach one settings object to multiple datasets with different distributions. 

This also makes it much easier to set up multi-phase tasks using different datasets.

**Review by burtenshaw — COMMENTED** [review]
This looks good. Some things to note.

## Docs/ Guide @sdiazlor @davidberenstein1957 @nataliaElv 
We need a detailed how to guide on this. Should we point it at this branch or do a separate PR?

I would cover these points in the guide:
- Define task distribution and show examples of effects of different settings
- Changing task distributions at different stages of the dataset lifecycle. Say what is possible at different points.
- Accessing and changing dataset setting with distributions
- Retrieving a dataset from argilla and changing its distribution settings

## Updating settings @frascuchon 
I think we need to revisit this when `rg.Settings` can be updated at more points in the lifecycle, and the settings UI is refined, to make SDK and UI settings consistent and logical.

File: argilla/tests/integration/test_create_datasets.py [hunk 8]
```diff
@@ -70,8 +68,7 @@ def test_create_multiple_dataset_with_same_settings(self, client: Argilla):
             assert schema["question"].name == "question"
             assert schema["question"].values == [1, 2, 3, 4, 5]
 
-    def test_create_dataset_from_existing_dataset(self, client: Argilla):
-        dataset_name = f"test_dataset_{uuid.uuid4()}"
+    def test_create_dataset_from_existing_dataset(self, client: Argilla, dataset_name: str):
```
Comment (davidberenstein1957) [inline argilla/tests/integration/test_create_datasets.py:71]: should we check if task_distribution is included correctly during such operations?

**Review by davidberenstein1957 — COMMENTED** [review]
LFTM, small comment w.r.t. tests. Also, I believe @nataliaElv starting to work on the docs with @sdiazlor.

File: argilla/tests/integration/test_create_datasets.py (hunk 8 above)
Comment (frascuchon) [reply argilla/tests/integration/test_create_datasets.py:71]: Done

File: .github/workflows/argilla.yml (hunk 2 above)
Comment (frascuchon) [inline .github/workflows/argilla.yml]:
```suggestion
        image: argilladev/argilla-quickstart:develop
```
